### DISCUSSION TITLE
Debugging with gds and gdb

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#1232] [Q&A] [answered]
Since there was a previous post regarding debugging without a clear conclusion (I had to do more tinkering after reading), I thought I'd just give my procedure for starting a deployment with GDS and gdb.

This works for F' 2.1.0 and possibly 3.0.

Assuming the `Ref` reference demo deployment, enter the deployment directory, and generate a unit-test build directory. Then, following @LeStarch 's instructions in [Set DebugInfo for help with debugging](https://github.com/nasa/fprime/discussions/918), build the main binary from the build directory. This process enables debugging symbols in the binary.
```
fprime-util generate --ut
cd build-fprime-automatic-native-ut
make Ref_exe
```

Now return to the deployment directory and run GDS with the debugging binary:
```
cd ..
fprime-gds --app ./build-fprime-automatic-native-ut/Linux/bin/Ref
```

To find the PID of the Ref binary:
```
ps -A | grep Ref
```

Open a new terminal, and run GDB, attaching to the correct process:
```
gdb --pid PID_FROM_PS
```

You can now set breakpoints, for example to break on entry to a telecommand handler. GDS is open in your browser, so you can send TCs and receive TMs like normal.

I was unable to run the deployment with gdb directly, since the deployment gave up lots of errors in the Framer, and GDS seemed unable to connect.

Feel free to add anything.

**Comment by JackNWhite** [comment, marked as answer]
Dummy answer

**Comment by timcanham** [comment]
That's great! Just be aware that if you connect to a thread that is being pinged by the Health component, it will eventually time out. You should turn of health checks with the `HLTH_ENABLE` command.

**Comment by T-recks** [comment]
I'm trying to get some debugging working for Ref using this thread and #918 but whenever I try the make Ref_exe step I get an error:

make: *** No rule to make target 'Ref_exe'.  Stop.

Any suggestions? Sorry I'm not too familiar with Cmake.

**Reply by thomas-bc** [reply]
This post is pretty old -- you can find the list of targets that are available in your build cache, under the CMakeFiles/ directory (e.g. `Ref/build-fprime-automatic-native-ut/CMakeFiles/`). If you're running FPrime 3, you're probably looking for the `Ref_ut_exe` target? But I would also recommend simply using the F' utility, e.g. `fprime-util build --ut` which should do what you're looking for.

**Reply by dheadrick1618** [reply]
Has this debug procedure changed since this last post? I am using Fprime v3.1.1. Generating and building the `build-fprime-automatic-native-ut` directory using the associated `fprime-util generate --ut` and `fprime-util build --ut` commands does not result in the creation of a deployment Ref_ut_exe binary. Instead all I can seem to find in 'Ref/build-fprime-automatic-native-ut/bin/Linux' is what seems to be individual component binaries. If I try running the gds on one of these individual binaries such as:` fprime-gds --app ./build-fprime-automatic-native-ut/bin/Linux/Ref_MathSender_ut_exe` , then tests are ran automatically and I get an error: 
`[INFO] Shutting down F prime due to error. Failed to run Ref_MathSender_ut_exe Application` .
